File: providers/aws/src/main/java/cloudspec/aws/ec2/loader/EC2VpcPeeringConnectionLoader.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2.loader;

import cloudspec.aws.IAWSClientsProvider;
import cloudspec.aws.ec2.resource.EC2VpcPeeringConnectionResource;
import software.amazon.awssdk.services.ec2.Ec2Client;
import software.amazon.awssdk.services.ec2.model.DescribeVpcPeeringConnectionsResponse;
import software.amazon.awssdk.utils.IoUtils;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class EC2VpcPeeringConnectionLoader extends EC2ResourceLoader<EC2VpcPeeringConnectionResource> {
    private final IAWSClientsProvider clientsProvider;

    public EC2VpcPeeringConnectionLoader(IAWSClientsProvider clientsProvider) {
        this.clientsProvider = clientsProvider;
    }

    @Override
    public Optional<EC2VpcPeeringConnectionResource> getById(String vpcPeeringConnectionId) {
        return getPeeringConnections(Collections.singletonList(vpcPeeringConnectionId)).findFirst();
    }

    @Override
    public List<EC2VpcPeeringConnectionResource> getAll() {
        return getPeeringConnections().collect(Collectors.toList());
    }

    private Stream<EC2VpcPeeringConnectionResource> getPeeringConnections() {
        return getPeeringConnections(Collections.emptyList());
    }

    private Stream<EC2VpcPeeringConnectionResource> getPeeringConnections(List<String> vpcPeeringConnectionIds) {
        Ec2Client ec2Client = clientsProvider.getEc2Client();

        try {
            return ec2Client.describeRegions()
                    .regions()
                    .stream()
                    .flatMap(region -> getPeeringConnectionsInRegion(region, vpcPeeringConnectionIds));
        } finally {
            IoUtils.closeQuietly(ec2Client, null);
        }
    }

    private Stream<EC2VpcPeeringConnectionResource> getPeeringConnectionsInRegion(software.amazon.awssdk.services.ec2.model.Region region,
                                                                                  List<String> vpcPeeringConnectionIds) {
        Ec2Client client = clientsProvider.getEc2ClientForRegion(region.regionName());

        try {
            DescribeVpcPeeringConnectionsResponse response = vpcPeeringConnectionIds != null && !vpcPeeringConnectionIds.isEmpty() ?
                    client.describeVpcPeeringConnections(builder -> builder.vpcPeeringConnectionIds(vpcPeeringConnectionIds.toArray(new String[0]))) :
                    client.describeVpcPeeringConnections();

            return response.vpcPeeringConnections()
                    .stream()
                    .map(vpcPeeringConnection -> EC2VpcPeeringConnectionResource.fromSdk(region.regionName(), vpcPeeringConnection));
        } finally {
            IoUtils.closeQuietly(client, null);
        }
    }
}

File: providers/aws/src/main/kotlin/cloudspec/aws/ec2/EC2VpcPeeringConnectionLoader.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2

import arrow.fx.IO
import cloudspec.aws.IAWSClientsProvider
import kotlin.streams.toList

class EC2VpcPeeringConnectionLoader(clientsProvider: IAWSClientsProvider) :
        EC2ResourceLoader<EC2VpcPeeringConnection>(clientsProvider) {

    override fun resourcesInRegion(region: String,
                                   ids: List<String>): IO<List<EC2VpcPeeringConnection>> =
            requestInRegion(region) { client ->
                val filters = buildFilters(
                        mapOf(
                                FILTER_VPC_PEERING_CONNECTION_ID to ids
                        )
                )

                client
                    // https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeVpcPeeringConnections.html
                    .describeVpcPeeringConnections { builder ->
                        if (filters.isNotEmpty()) {
                            builder.filters(filters)
                        }
                    }
                    .vpcPeeringConnections()
                    .stream()
                    .map { it.toEC2VpcPeeringConnection(region) }
                    .toList()
            }

    companion object {
        private const val FILTER_VPC_PEERING_CONNECTION_ID = "vpc-peering-connection-id"
    }
}
